diffrax/integrate.py

```diff
@@ -38,6 +40,77 @@
 from .term import AbstractTerm, VectorFieldWrapper, WrapTerm
 
 
+class _HistoryVectorField(eqx.Module):
+    vector_field: Callable
+    t0: float
+    tprev: float
+    tnext: float
+    dense_info: PyTree[Array]
+    dense_interp: Optional[DenseInterpolation]
+    interpolation_cls: Type[AbstractGlobalInterpolation]
+    y0_history: Callable
+    delays: PyTree[Callable]
+
+    def __call__(self, t, y, args):
+        history_vals = []
+        delays, treedef = jtu.tree_flatten(self.delays)
+        if self.dense_interp is None:
+            for delay in self.delays:
+                delay_val = delay(t, y, args)
+                alpha_val = t - delay_val
+                y0_val = self.y0_history(alpha_val)
+                history_vals.append(y0_val)
+        else:
+            for delay in delays:
+                delay_val = delay(t, y, args)
+                alpha_val = t - delay_val
+
+                is_before_t0 = alpha_val < self.t0
+                is_before_tprev = alpha_val < self.tprev
+                at_most_t0 = jnp.where(alpha_val < self.t0, alpha_val, self.t0)
+                t0_to_tprev = jnp.clip(alpha_val, self.t0, self.tprev)
+                at_least_tprev = jnp.where(
+                    alpha_val < self.tprev, self.tprev, alpha_val
+                )
+
+                if self.dense_info is None:
+                    history_val = jnp.where(
+                        is_before_t0,
+                        self.y0_history(at_most_t0),
+                        self.dense_interp.evaluate(t0_to_tprev),
+                    )
```

Comment:
Is this branch ever taken? I think either both `dense_interp` and `dense_info` are `None`, or neither of them are?
It would be neater to have only a single branch here, and to enforce this invariant with `assert` statements:

```python
if self.dense_interp is None:
    assert self.dense_info is None
    ...
else:
    assert self.dense_info is not None
    ...
```

Reply:
The branch was taken during the init of `diffeqsolve` but this is indeed cleaner way of doing it !